docs/dev/datastream_execution_mode.md

```diff
@@ -0,0 +1,246 @@
+low-latency processing.
+
+#### BATCH Execution Mode
+
+In BATCH execution mode, the tasks of a job can be separated into stages that
+can be executed one after another. We can do this because the input is bounded
+and Flink can therefore fully process one stage of the pipeline before moving
+on to the next. In the above example the job would have three stages that
+correspond to the three tasks that are separated by the shuffle barriers.
+
+Instead of sending records immediately to downstream tasks, as explained above
+for STREAMING mode, processing in stages requires Flink to materialize
+intermediate results of tasks to some non-ephemeral storage which allows
+downstream tasks to read them after upstream tasks have already gone off line.
+This will increase the latency of processing but comes with other interesting
+properties. For one, this allows Flink to backtrack to the latest available
+results when a failure happens instead of restarting the whole job. Another
+side effect is that BATCH jobs can execute on fewer resources (in terms of
+available slots at TaskManagers) because the system can execute tasks
+sequentially one after the other.
+
+TaskManagers will keep intermediate results at least as long as downstream
+tasks have not consumed them. After that, they will be kept for as long as
+space allows in order to allow the aforementioned backtracking to earlier
+results in case of a failure.
+
+### State Backends / State
+
+- mention that in BATCH mode a special state backend is used and not the
+  configured one
+
+### Event Time / Watermarks
+
+### Processing Time
+
+### Failure Recovery
+
+In STREAMING execution mode, Flink uses checkpoints for failure recovery. Take a
+look at the [checkpointing documentation]({% link
+dev/stream/state/checkpointing.md %}) for hands-on documentation about this and
+how to configure it. There is also a more introductory section about [fault
+tolerance via state snapshots]({% link learn-flink/fault_tolerance.md %}) that
+explains the concepts at a higher level.
+
+One of the characteristics of checkpointing for failure recovery is that Flink
+will restart all the running tasks from a checkpoint in case of a failure. This
+can be more costly than what we have to do in BATCH mode (as explained below),
+which is one of the reasons that you should use BATCH execution mode if your
+job allows it.
+
+In BATCH execution mode, Flink will try and backtrack to previous processing
+stages for which intermediate results are still available. Potentially, only
+the tasks that failed (or their predecessors in the graph) will have to be
+restarted, which can improve processing efficiency and overall processing time
+of the job compared to restarting all tasks from a checkpoint.
+
+## Important Considerations
+
+What doesn't work:
+ - broadcast state/pattern
+ - iterations
+ - operations that rely on checkpointing
+ - this includes most "regular" sinks
```

Comment:
```suggestion
Full unification of batch and stream processing under the DataStream API is still under active development. 
That means there are still a number of DataStream features that only work under the `STREAMING` execution mode. 

 - broadcast state/pattern
 - iterations
 - operations that rely on checkpointing
 - this includes most "regular" sinks
```

Aljoscha, I'll need you to expand on the bullet points.

Reply:
Sorry, I forgot to mention that this documentation is WIP and that Dawid has some incoming changes for this part. 🙈

Reply:
Sorry, Teach